.bazelrc

```diff
@@ -1,47 +1,57 @@
 startup --output_base=/tmp/bazel
 
+# STYLE CONFIGS
 build --announce_rc
 build --color=yes
+
+# C/C++ CONFIGS
 build --cxxopt=-std=c++14
 build --strip=never
 build --compilation_mode=dbg
 
-# Needed for coverage collection
+# MAGMA VM CONFIGS
+build:specify_vm_cc --action_env=CC=/usr/bin/gcc
+build:specify_vm_cc --action_env=CXX=/usr/bin/g++
+build:vm --disk_cache=/home/vagrant/magma/.bazel-cache
+common:vm --repository_cache=/home/vagrant/magma/.bazel-cache-repo
+build:vm --define=folly_so=1
+build:vm --config=specify_vm_cc
+
+# MAGMA-BUILDER DOCKER CONTAINER CONFIGS
+build:docker --disk_cache=/magma/.bazel-cache
+common:docker --repository_cache=/magma/.bazel-cache-repo
+
+# DEVCONTAINER CONFIGS
+build:devcontainer --disk_cache=/workspaces/magma/.bazel-cache
+common:devcontainer --repository_cache=/workspaces/magma/.bazel-cache-repo
+build:devcontainer --define=folly_so=1
+
+# CODE COVERAGE CONFIGS
 build --javacopt="-source 8"
 build --javacopt="-target 8"
 build --host_javabase=@bazel_tools//tools/jdk:remote_jdk11
 build --javabase=@bazel_tools//tools/jdk:remote_jdk11
 build --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_java11
 build --java_toolchain=@bazel_tools//tools/jdk:toolchain_java11
 
-# Coverage options
 coverage --config=coverage
 build:coverage --combined_report=lcov
 build:coverage --instrumentation_filter="//(orc8r|lte)/gateway/c[/:],-//(orc8r|lte)/protos[/:],-/*/test[/:]"
 
-# For building with Bazel inside the Magma-Builder Docker container
-build:docker --disk_cache=/magma/.bazel-cache
-common:docker --repository_cache=/magma/.bazel-cache-repo
-
-build:devcontainer --disk_cache=/workspaces/magma/.bazel-cache
-common:devcontainer --repository_cache=/workspaces/magma/.bazel-cache-repo
-build:devcontainer --define=folly_so=1
-
-build:specify_vm_cc --action_env=CC=/usr/bin/gcc
-build:specify_vm_cc --action_env=CXX=/usr/bin/g++
-
-# For building with Bazel inside the Magma VM
-build:vm --disk_cache=/home/vagrant/magma/.bazel-cache
-common:vm --repository_cache=/home/vagrant/magma/.bazel-cache-repo
-build:vm --define=folly_so=1
-build:vm --config=specify_vm_cc
-
-build:asan --strip=never
+# ASAN
 build:asan --copt=-fsanitize=address
 build:asan --copt=-O0
 build:asan --copt=-fno-omit-frame-pointer
 build:asan --linkopt=-fsanitize=address
 build:asan --action_env=ASAN_OPTIONS=detect_leaks=1:color=always
+build:asan --test_tag_filters=-no_asan
+
+# LSAN (Enabled by default)
```

Comment:
I think it's worth discussing whether this is enabled by default or not.

In my prior life, we separated the various sanitizers and required they be specified with `--config` clauses. I haven't stopped to think about why, but here are some ideas (not all of them strong):

- Most sanitizers cannot be run simultaneously (except LSAN, ASAN?), so having one default to one is a little misleading IMO
- I wonder (but don't know) if test runtime behavior is sometimes modified by `LSAN` presence (at a minimum timings? memory use...)
- Failure of a CI job that clearly specifies it is an `ASAN unit test run` provides improved signal at-a-glance as to root cause (e.g. you note that non-ASAN is fine, ASAN failing -> better signal on problem at-a-glance)

Upshot: I would advocate (but welcome discussion) that we do not default to ASAN-on / LSAN on.

Downsides that I can think of:

- Longer build / test cycles (have to build multiple unit test variants and run them)
- Contributor has to know there are `--config=` options that need to be tested aside from vanilla `bazel test //to/target` (but they will learn thanks to CI...)

Reply:
yeah that is a good point. I was mostly defaulting to the current behavior, which is ASAN for debug builds, LSAN for production builds. I definitely think we don't want to enable ASAN by default, but rather have a CI check that runs all tests with ASAN.

Reply:
Whoops - I misread the PR (actually, perhaps minimally read the actual diffs) - sorry about that. I saw preserve existing behavior and knew that debug builds in CMake use ASAN and thought: would prefer not!

Reply:
Aside but we also tended to run all unit tests in release mode as well, as part of CI. So you end up with debug + release mode, and in DEBUG modes you cross product with {ASAN, MSAN, TSAN, and on}. LOTS of tests =D

Reply:
Yeah... and changing these configs trigger a re-build in all the external dependencies as well. Which can be slow :/ I'll do a quick look around in sourcegraph to see how people test these configs. :D

Reply:
> For LSAN, I do not know too much about the side effects yet to see if keeping them enabled by default is a good / bad idea.

Me neither, but I generally hesitate as to whether this is best practice for production builds (or again, whether bundling with default build behavior is the right call).  But if we don't people may forget the `--config=lsan` and produce production binaries not what they expect (though a human should not be running these commands...)

Reply:
yeah, definitely agreed about limiting implicit default configs. Maybe I will keep lsan not enabled by default, but wrap it inside like a `make build_prod` Make command just to keep them explicit.

Reply:
modified PR to not enable it by default.

Reply:
I am starting to wonder if having a project wide lsan/asan configuration is not the way to go as it affects external dependency builds? Maybe we can create a custom rule that handles ASAN/LSAN configurations for Magma C/C++ targets...

envoyproxy does something like this, but they are configuring a lot more than what I'm proposing.https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/bazel/envoy_library.bzl?L84

Reply:
> I am starting to wonder if having a project wide lsan/asan configuration is not the way to go as it affects external dependency builds?

Interesting. Does seem like we are less likely to need lsan/asan coverage for third party build targets being statically linked... I'm personally unaware if there are any gotchas in not having total lsan/asan coverage for all statically linked assets in a binary? Would want to know that.  E.g. not havinbg lsan/asan on for Folly - does that in any way hinder lsan/asan behavior at the use of folly lib (can it potentially mask issues in our code).

Reply:
Yeah I was wondering if this would mask issues as well. Will google around. I was also wondering if we could use the custom rule concept to have different levels of compiler warnings for external libraries / internal. (it's just a little noisy for some libraries right now) But not clear on whether that is good practice or not.